Features!

Grissess opened the pull request:
- !quotedel to remove quotes
- Safer file I/O, more Pythonic syntax
- Actual shuffling to avoid repeating quotes for specific patterns
- Using SystemRandom for cryptographically random quotes ;)
- Config: remove-votes for voters to do a !quotedel
- Config: max-history for max number of !quote shuffles to remember

sjrct commented:
Can you change the first line of your commit message to something slightly more descriptive?

Maybe "Add !quotedel; safer file I/O; improve randomness" or something else that briefly summarizes the changes.

sjrct commented on quote.py:
```diff
@@ -3,48 +3,102 @@
-	quotes = []
+	# Configuration defaults
+	remove_votes = 3
+	max_history = 10
```
I would just have the defaults be defined in the 2nd parameter of the `config.get` below. Seems simpler.

Grissess replied:
Certainly not a problem to do it this way, but I like the verbosity/clarity this approach provides. The redundancy is a slight cost, of course :P

sjrct replied:
Fair point.

richteer replied:
Agreed with @sjrct, but keep the deletion of the `quotes = []`, that shouldn't be there anymore anyway.

Config defaults will be seeing a change soon from core, so don't worry about that.

EDIT: Github is :hankey: leaving this comment anyway

sjrct commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+				with open(self.path, 'r') as f:
+					self.quotes.extend([x.rstrip() for x in f if x.rstrip()])
+			except IOError, OSError:
+				pass
```
We should probably log the error message.

Grissess replied:
print() acceptable in this environment?

sjrct replied:
`self.log.error` is preferable.

sjrct commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+					self.rand_quotes.remove(sr.choice(self.rand_quotes))  # XXX LRU?
+
+			self.rand_quotes[pattern] = ls
+			quote = self.quote(pattern)  # Recursion!
```
Why don't we just move the True block of the if statement outside and after the if and remove the recursive call?

Grissess replied:
Because I clearly needed more coffee :P 

Will fix...

sjrct replied:
That's why I take my coffee by IV drip.

richteer replied:
+1 to coffee IV

sjrct commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+			if self.path is not None:
+				try:
+					with open(self.path, 'r+') as f:
+						f.writelines([q for q in f if q.rstrip() != quote])
```
Should we perhaps just write out `self.quotes` here instead?

Grissess replied:
Yes, fixing...

sjrct reviewed (commented):
There is also the problem that the voting mechanism can be easily circumvented now by simply changing you nickname a bunch of times. Halibot support for authenticated/true names is not there yet though, so this is not exactly fixable with the current halibot. Perhaps we should add the ability to disable `!quotedel` to the config though, because of this.

richteer commented on quote.py:
```diff
@@ -3,48 +3,102 @@
 import random
 from halibot import HalModule
 
+sr = random.SystemRandom()
```
Is this needed? Docs say this isn't necessarily implemented on all platforms, which gets a NACK from me.

Grissess replied:
Is it not? Checking the docs...

sjrct replied:
https://docs.python.org/3/library/random.html#random.SystemRandom

Grissess replied:
...eh, you're right, but they're not specific as to what systems might not support it--while it seems rare, I suppose relying on the pseudorandom internal generator is a better choice for ultimate portability. Fixing...

richteer commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+				try:
+					expr = re.compile(pattern)
+				except re.error as e:
+					return 'Invalid pattern: %s' % (e,)
```
Please use `.format()` instead of the `%` syntax. Only marking this one, but applies to the other ones too

Grissess replied:
For a second, I wasn't sure if this was Python 3, so I was being as close to Python 2 spec as possible :P 

If you don't mind me dropping Py2 compat, I'll gladly do that right now :D

richteer replied:
There is no py2 compat :)

We're using all the asyncio stuff, which afaik is like, 3.2+ or something

Grissess replied:
I was a dumb, fixed anyway :P

richteer commented on quote.py:
```diff
@@ -3,48 +3,102 @@
 		return 'Added :)'
 
+	def quotedel(self, quote, nick):
+		if quote not in self.quotes:
```
Maybe use a regex, and require it to match exactly one quote? Should make the voting slightly easier for those without the copy/paste.

Grissess replied:
Working on that...

sjrct replied:
I could see this being made an issue and fixed later, though.

richteer replied:
Thought on this a little longer, might be useful to print a useful message as confirmation, either that it matched too many quotes, or a truncated version of the quote to confirm.

Grissess replied:
I like the idea of printing out a part of it to confirm, that'd help prevent false positives :P

I might also print out a clipped version of the start and end of a quote for multiple matches, up to a certain number and certain length. But I'm wary that a large RE and database can cause this to have some problems, so I'm going to try and keep it as reasonable as possible.

Also, these comments refer to blobs that probably don't exist since the last push -f :P

richteer commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+				with open(self.path, 'r') as f:
+					self.quotes.extend([x.rstrip() for x in f if x.rstrip()])
+			except IOError, OSError:
+				pass
```
Each module gets a `self.log` logger instance, can just use that

richteer reviewed (changes requested):
Need some tweaks, but I like the general ideas. Feel free to argue any of the points :)

Grissess commented:
Working on it right now, in fact :P 

@sjrct I am aware of the ability to do multiple votes, which is why I consulted with @cmr about that problem--he said it didn't seem like a problem for our use case, but seeing that it can be a problem elsewhere, I will certainly make it switchable in the config. (Technically, you can functionally do that right now by setting remove-votes to a ridiculously impossible number, but that's not particularly ergonomic :P) If you have any ideas for a way to more properly avert vote fraud, I'd be glad to hear them :D 

Changes pushed—let me know what you think!

sjrct commented on quote.py:
```diff
@@ -3,48 +3,102 @@
+					self.rand_quotes.remove(sr.choice(self.rand_quotes))  # XXX LRU?
+
+			self.rand_quotes[pattern] = ls
+			quote = self.quote(pattern)  # Recursion!
```
Actually perhaps my drip is dripping to slowly, this change actually breaks the no-pattern case. :sob: 

I don't mind how it is fixed, could put it back to the way it was, or early return, or set the `ls` inside the true case of the if and not at the end of the false case.

sjrct commented:
@Grissess, in the future we should allow only people who have authed with the server to vote, although that is still permissive of some vote fraud, it should diminish it quite a bit. When the admin module is written (I'm looking at you @richteer :angry:), we should be able to limit voting to only an approved group of authed users. But for now, there is not much that can be done.

Grissess commented:
@sjrct Looking forward to an auth module...how might one refer to an authorized group or holders of a permission? It's good to be prepared for that change in structure :P

sjrct commented:
@Grissess, that has yet to be precisely determined, I have an idea or two I like, but I think richteer has different ones. We will probably touch on it at the next meeting. I was thinking something along the lines of `message.author.hasPermission(role)` or something like that.

sjrct reviewed (approved):
Looks good to me. Visual tested only.

richteer commented:
@sjrct: Something like a `hasPermission` switch is what I'm thinking too, just not sure on the backend. Also decorators that do that too, because I hate readable code :)

richteer commented on quote.py:
```diff
@@ -1,50 +1,118 @@
+			try:
+				with open(self.path, 'r') as f:
+					self.quotes.extend([x.rstrip() for x in f if x.rstrip()])
+			except IOError, OSError:
```
Syntax error here, these need to be a tuple apparently.

Grissess replied:
Right, Python 3, yadda yadda, gimme a sec

richteer reviewed (changes requested):
Looks better, but `!quote` doesn't return any quotes now. Please test this if you can get the chance (feel free to use ##haltesting, check the readme.md for a sample config)

Grissess commented:
@richteer slightly better? :P 

Oh, I see: I suppose I should actually fire up the testing rig that I put...somewhere.

One moment...

Grissess committed:
Deletion; better randomization

- Now tracks patterns and pulls repeated invocations from a cached
  shuffle()
- Deletion can be done by reaching a vote threshold
- Config: max-history: controls the number of cached shuffles
- Config: remove-votes: how many votes are needed to remove a quote
- Config: can-remove: whether or not removal is enabled
